unixpwd/c/unixpwd_stubs.c

```diff
@@ -20,104 +20,90 @@
 {
     CAMLparam1(caml_user);
-    const char     *user;
-    char           *passwd;
+    char     *user;
+    char     *passwd;
+    int       saved_errno;
     CAMLlocal1(pw);
 
-    user = String_val(caml_user);
-    passwd = unixpwd_getpwd(user);
-    if (passwd == NULL && errno != 0)
-        caml_failwith(strerror(errno));
-    if (passwd == NULL)
-        caml_failwith("unspecified error in caml_unixpwd_getpwd()");
+    user = caml_stat_strdup(String_val(caml_user));
+    caml_enter_blocking_section();
+    errno = 0;
+    passwd = f(user);
+    saved_errno = errno;
+    caml_stat_free(user); user = NULL;
+    caml_leave_blocking_section();
+    errno = saved_errno;
+
+    if (passwd == NULL) /* errno of 0 will be mapped to `EUNKNOWNERR of 0` */
+        uerror(fname, caml_user);
 
     pw = caml_copy_string(passwd);
     free(passwd);
     CAMLreturn(pw);
 }
 
 CAMLprim        value
-caml_unixpwd_getspw(value caml_user)
+caml_unixpwd_getpwd(value caml_user)
 {
-    CAMLparam1(caml_user);
-    const char     *user;
-    char           *passwd;
-    CAMLlocal1(pw);
-
-    user = String_val(caml_user);
-    passwd = unixpwd_getspw(user);
-    if (passwd == NULL && errno != 0)
-        caml_failwith(strerror(errno));
-    if (passwd == NULL)
-        caml_failwith("unspecified error in caml_unixpwd_getspw()");
-
-    pw = caml_copy_string(passwd);
-    free(passwd);
-    CAMLreturn(pw);
+    return caml_unixpwd_get_(caml_user, "unixpwd_getpwd", unixpwd_getpwd);
 }
 
-
+CAMLprim        value
+caml_unixpwd_getspw(value caml_user)
+{
+    return caml_unixpwd_get_(caml_user, "unixpwd_getspw", unixpwd_getspw);
+}
 
 CAMLprim        value
 caml_unixpwd_get(value caml_user)
 {
-    CAMLparam1(caml_user);
-    const char     *user;
-    char           *passwd;
-    CAMLlocal1(pw);
-
-    user = String_val(caml_user);
-    passwd = unixpwd_get(user);
-    if (passwd == NULL && errno != 0)
-        caml_failwith(strerror(errno));
-    if (passwd == NULL)
-        caml_failwith("unspecified error in caml_unixpwd_get()");
-
-    pw = caml_copy_string(passwd);
-    free(passwd);
-    CAMLreturn(pw);
+    return caml_unixpwd_get_(caml_user, "unixpwd_get", unixpwd_get);
 }
 
-CAMLprim        value
-caml_unixpwd_setpwd(value caml_user, value caml_password)
+static CAMLprim value caml_unixpwd_set_(value caml_user, value caml_password, const char *fname, int(*f)(const char*, char*))
 {
     CAMLparam2(caml_user, caml_password);
-    const char     *user;
-    char           *password;
-    int             rc;
+    char     *user;
+    char     *password;
+    int       saved_errno;
+    int       rc;
 
-    user = String_val(caml_user);
+    user = caml_stat_strdup(String_val(caml_user));
     password = caml_stat_strdup(String_val(caml_password));
-    rc = unixpwd_setpwd(user, password);
+    caml_enter_blocking_section();
+    errno = 0;
+    rc = f(user, password);
+    saved_errno = errno;
+    caml_stat_free(user);
```

Comment:
Again, it appears to me that memory ownership is subtle.

Reply:
Lets keep the `caml_stat*` functions in the section where the runtime lock is held for consistency with the allocation function (where the runtime lock is held).